File: collect_app/src/androidTest/java/org/odk/collect/android/support/async/AsyncWorkTrackerIdlingResource.java
package org.odk.collect.android.support.async;

import androidx.test.espresso.IdlingResource;

public class AsyncWorkTrackerIdlingResource implements IdlingResource {

    private ResourceCallback resourceCallback;

    @Override
    public String getName() {
        return AsyncWorkTrackerIdlingResource.class.getName();
    }

    @Override
    public boolean isIdleNow() {
        boolean idle = AsyncWorkTracker.getTaskCount() == 0;
        if (idle && resourceCallback != null) {
            resourceCallback.onTransitionToIdle();
        }

        return idle;
    }

    @Override
    public void registerIdleTransitionCallback(ResourceCallback resourceCallback) {
        this.resourceCallback = resourceCallback;
    }
}

File: collect_app/src/androidTest/java/org/odk/collect/android/support/async/AsyncWorkTrackerIdlingResource.kt
package org.odk.collect.android.support.async

import androidx.test.espresso.IdlingResource
import org.odk.collect.android.support.async.AsyncWorkTracker.taskCount

class AsyncWorkTrackerIdlingResource : IdlingResource {

    private var resourceCallback: IdlingResource.ResourceCallback? = null

    override fun getName(): String {
        return AsyncWorkTrackerIdlingResource::class.java.name
    }

    override fun isIdleNow(): Boolean {
        val idle = taskCount == 0
        if (idle) {
            resourceCallback?.onTransitionToIdle()
        }

        return idle
    }

    override fun registerIdleTransitionCallback(resourceCallback: IdlingResource.ResourceCallback?) {
        this.resourceCallback = resourceCallback
    }
}
